packages/shared/src/time_control/time_control_systems.ts

```diff
@@ -0,0 +1,258 @@
+
+export interface TimeControl<ConfigT, StateT> {
+  initialState(config: ConfigT): StateT;
+  /** Return an updated state to reflect the elapsed time */
+  elapse(elapsedMS: number, state: StateT, config: ConfigT): StateT;
+  /** Renew period time in systems like Byo-Yomi or Simple. */
+  renew(state: StateT, config: ConfigT): StateT;
+  msUntilTimeout(state: StateT, config: ConfigT): number;
+  timeString(state: StateT): string;
+}
+
+export interface IBasicTimeState {
+  remainingTimeMS: number;
+}
+
+class FischerTimeControl
+  implements TimeControl<IFischerConfig, IBasicTimeState>
+{
+  initialState(config: IFischerConfig): IBasicTimeState {
+    return { remainingTimeMS: config.mainTimeMS };
+  }
+
+  elapse(elapsedMS: number, state: IBasicTimeState) {
+    return { remainingTimeMS: state.remainingTimeMS - elapsedMS };
+  }
+
+  renew(state: IBasicTimeState, config: IFischerConfig) {
+    const uncapped = state.remainingTimeMS + config.incrementMS;
+
+    const remainingTimeMS =
+      config.maxTimeMS === null
+        ? uncapped
+        : Math.min(uncapped, config.maxTimeMS);
+
+    return { remainingTimeMS };
+  }
+
+  msUntilTimeout(state: IBasicTimeState): number {
+    return state.remainingTimeMS;
+  }
+
+  timeString(state: IBasicTimeState): string {
+    return msToTime(state.remainingTimeMS);
+  }
+}
+
+class AbsoluteTimeControl
+  implements TimeControl<ITimeControlConfig, IBasicTimeState>
+{
+  initialState(config: ITimeControlConfig): IBasicTimeState {
+    return { remainingTimeMS: config.mainTimeMS };
+  }
+
+  elapse(elapsedMS: number, state: IBasicTimeState) {
+    return { remainingTimeMS: state.remainingTimeMS - elapsedMS };
+  }
+
+  renew = nullRenewal;
+
+  msUntilTimeout(state: IBasicTimeState): number {
+    return state.remainingTimeMS;
+  }
+
+  timeString(state: IBasicTimeState): string {
+    return msToTime(state.remainingTimeMS);
+  }
+}
+
+export interface IByoYomiConfig extends ITimeControlConfig {
+  type: TimeControlType.ByoYomi;
+  numPeriods: number;
+  periodTimeMS: number;
+}
+export interface IByoYomiState {
+  mainTimeRemainingMS: number;
+  periodsRemaining: number;
+  periodTimeRemainingMS: number;
+}
+
+class ByoYomiTimeControl implements TimeControl<IByoYomiConfig, IByoYomiState> {
+  initialState(config: IByoYomiConfig): IByoYomiState {
+    return {
+      mainTimeRemainingMS: config.mainTimeMS,
+      periodsRemaining: config.numPeriods,
+      periodTimeRemainingMS: config.periodTimeMS,
+    };
+  }
+
+  elapse(
+    elapsedMS: number,
+    state: IByoYomiState,
+    config: IByoYomiConfig,
+  ): IByoYomiState {
+    state = { ...state };
+    if (state.mainTimeRemainingMS) {
+      if (state.mainTimeRemainingMS >= elapsedMS) {
+        state.mainTimeRemainingMS -= elapsedMS;
+        return state;
+      }
+      elapsedMS -= state.mainTimeRemainingMS;
+      state.mainTimeRemainingMS = 0;
+    }
+    while (state.periodTimeRemainingMS < elapsedMS) {
```

Comment:
My instinct would be to replace this while by division with remainder. Something like
```
let periodsUsed = 0;

// first period
if (state.periodTimeRemainingMS < elapsedMS) {
  elapsedMS -= state.periodTimeRemainingMS;
  periodsUsed += 1;
}

let remainder = elapsedMS % config.periodTimeMS;
const q = (elapsedMS - remainder) / config.periodTimeMS; // should give integer result
periodsUsed += q;

state.numPeriods = Math.max(0, state.numPeriods - periodsUsed);
state.periodTimeRemainingMS = state.numPeriods === 0
  ? 0
  : config.periodTimeMS - remainder;
```
I haven't tried this out yet, and actually, I suspect your solution has better performance (as long as there are not many periods used in one move, which can't happen often anyways).